### PR TITLE
Add secret scanning to SCM guide, fixes #488

### LINKED ISSUE #488: Add secret scanning and push protection to SCM-BestPractices recommendations
It's sadly easy to accidentally insert secrets into a repository ([here's an example](https://docs.google.com/document/d/1MWBTqpO8XofvN9ElTX5tPB8a7N1N0Wp9JgEQ9ff4Qvo/edit)).

We should modify the [SCM Best Practices](https://best.openssf.org/SCM-BestPractices/#recommendations) to say that any SCM should (where practical) enable scanning for secrets in a repo (including in proposed merge requests / pull requests), and then warn/prevent them (unless specially approved). E.g., in GitHub, secret scanning and push protection should be enabled. Linux Foundation projects can use LFX to use another secret scanning tool.

Related: https://github.com/ossf/tac/issues/215 

When implementing this, Set both as defaults for new projects, and add scanning to existing projects. Once the scanning for existing projects looks okay, add push protection.

### PR DESCRIPTION
**Pull request by david-a-wheeler**

### CODE REVIEW

**Comment by david-a-wheeler** [comment]
@SecurityCRob - comments?

I did *not* try to edit `best-practices.yml`. I'm not sure what that yml file is doing. Is that the source & the README is generated? I don't see any code to do the generation. If the .yml file is generated from the README, then please run that tool. If they're edited simultaneously, well, yuck :-(. In any case, there should be a clear document *somewhere* what their relationship is.

File: docs/SCM-BestPractices/README.md [hunk 1]
```diff
@@ -126,6 +126,7 @@ For recommendations only applicable to GitHub or GitLab visit one of the followi
...
 * Webhook Configured Without SSL Verification [<img src="https://user-images.githubusercontent.com/287526/230376963-ae9b8a47-4a74-4746-bc83-5b34cc520d40.svg" alt="GitLab" height="20" width="20">](gitlab/project/project_webhook_doesnt_require_ssl.md) [GitLab](gitlab/project/project_webhook_doesnt_require_ssl.md)
 * Project Should Have Fewer Than Three Owners [<img src="https://user-images.githubusercontent.com/287526/230376963-ae9b8a47-4a74-4746-bc83-5b34cc520d40.svg" alt="GitLab" height="20" width="20">](gitlab/project/project_has_too_many_admins.md) [GitLab](gitlab/project/project_has_too_many_admins.md)
+* Secret Scanning Should be Enabled [<img src="https://user-images.githubusercontent.com/287526/230375178-2f1f8844-5609-4ef3-b9ac-141c20c43406.svg" alt="GitHub" height="20" width="20">](github/repository/forking_allowed_for_repository.md) [GitHub](github/repository/secret_scanning.md)
```
Comment (ctcpip) [inline docs/SCM-BestPractices/README.md:129]:
```suggestion
* Secret Scanning Should be Enabled [<img src="https://user-images.githubusercontent.com/287526/230375178-2f1f8844-5609-4ef3-b9ac-141c20c43406.svg" alt="GitHub" height="20" width="20">](github/repository/secret_scanning.md) [GitHub](github/repository/secret_scanning.md)
```

File: docs/SCM-BestPractices/github/repository/secret_scanning.md [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+severity: HIGH
+
+## Description
+
+This security control checks for secrets being checked into a repository
+and prevents it.
+Turning on secret scanning detects many kinds of secrets being checked
+into a repository and reports it.
+Push protection builds on secret scanning and prevents attempts to add such
+secrets (and creates and alert if that it overridden).
```
Comment (ctcpip) [inline docs/SCM-BestPractices/github/repository/secret_scanning.md:14]:
```suggestion
secrets (and creates an alert if that is overridden).
```

File: docs/SCM-BestPractices/github/repository/secret_scanning.md [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+
+Inserting a secret into a source code repository is unfortunately an easy
+mistake to make. Instructions for various services often
+encourage putting secrets into a repository (as it's the "simple" way
+to do it, though not the secure way).
+Once secrets are in a repository, they become available for anyone who
+can view the repository (directly or via a copy).
+Inserting secrets into a source code repository is a mistake, since
+it interferes with the whole point of a secret: keeping it secret.
+This is such a common kind of vulnerability that it is identified as
```
Comment (ctcpip) [inline docs/SCM-BestPractices/github/repository/secret_scanning.md:26]:
```suggestion
This is such a common type of vulnerability that it is identified as
```

File: docs/SCM-BestPractices/github/repository/secret_scanning.md [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+
+### Enabling Secret Scanning
+
+[The GitHub documentation describes how to enable secret scanning](https://docs.github.com/en/code-security/secret-scanning/configuring-secret-scanning-for-your-repositories).
+
+In short, to enable GitHub secret scanning on a repository:
+
+1. Make sure you have admin permissions
+2. Go to the repo's settings page
+3. Enter "Security" section of the sidebar, click Code security and analysis.
```
Comment (ctcpip) [inline docs/SCM-BestPractices/github/repository/secret_scanning.md:50]:
```suggestion
3. Enter "Security" section of the sidebar, click "Code security and analysis".
```

File: docs/SCM-BestPractices/github/repository/secret_scanning.md [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+from pushing secrets to a repository and generates an alert
+whenever a contributor bypasses the block."
+
+[The GitHub documentation describes how to enable push protection with secret scanning](https://docs.github.com/en/code-security/secret-scanning/push-protection-for-repositories-and-organizations).
+
+In short, to enforce GitHub secret scanning on a repository:
+
+1. Make sure you have admin permissions
+2. Go to the repo's settings page
+3. Enter "Security" section of the sidebar, click Code security and analysis.
```
Comment (ctcpip) [inline docs/SCM-BestPractices/github/repository/secret_scanning.md:72]:
```suggestion
3. Enter "Security" section of the sidebar, click "Code security and analysis".
```

File: docs/SCM-BestPractices/github/repository/secret_scanning.md [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+4. Under "Code security and analysis", find "GitHub Advanced Security."
+5. Under "Secret scanning", under "Push protection", click Enable.
```
Comment (ctcpip) [inline docs/SCM-BestPractices/github/repository/secret_scanning.md:74]:
```suggestion
5. Under "Secret scanning", under "Push protection", click "Enable".
```

File: docs/SCM-BestPractices/gitlab/project/secret_scanning.md [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+Inserting a secret into a source code repository is unfortunately an easy
+mistake to make. Instructions for various services often
+encourage putting secrets into a repository (as it's the "simple" way
+to do it, though not the secure way).
+Once secrets are in a repository, they become available for anyone who
+can view the repository (directly or via a copy).
+Inserting secrets into a source code repository is a mistake, since
+it interferes with the whole point of a secret: keeping it secret.
+This is such a common kind of vulnerability that it is identified as
```
Comment (ctcpip) [inline docs/SCM-BestPractices/gitlab/project/secret_scanning.md:25]:
```suggestion
This is such a common type of vulnerability that it is identified as
```

**Review by ctcpip — CHANGES_REQUESTED** [review]
this is great!  looks good, just some minor corrections, and a few consistency suggestions as well

**Review by SecurityCRob — APPROVED** [review]
I like the addition.  If we could get Chris' open comment resolved I think we can move this forward

**Comment by SecurityCRob** [comment]
> @SecurityCRob - comments?
> 
> I did _not_ try to edit `best-practices.yml`. I'm not sure what that yml file is doing. Is that the source & the README is generated? I don't see any code to do the generation. If the .yml file is generated from the README, then please run that tool. If they're edited simultaneously, well, yuck :-(. In any case, there should be a clear document _somewhere_ what their relationship is.

I *think* Randall put that in place.  I'm also unsure how that works.  We should ask him to start.

**Comment by balteravishay** [comment]
> > @SecurityCRob - comments?
> > I did _not_ try to edit `best-practices.yml`. I'm not sure what that yml file is doing. Is that the source & the README is generated? I don't see any code to do the generation. If the .yml file is generated from the README, then please run that tool. If they're edited simultaneously, well, yuck :-(. In any case, there should be a clear document _somewhere_ what their relationship is.
> 
> I _think_ Randall put that in place. I'm also unsure how that works. We should ask him to start.

I think that might relate to how the [Legitify](https://github.com/Legit-Labs/legitify) was used to produce these recommendations. maybe @noamd-legit  can comment on this and suggest how to add new ones?

**Comment by david-a-wheeler** [comment]
In that case, I suggest removing the .yml file (in a separate pull request). We generally don't want generated files in the source repo.

@noamd-legit - would you mind creating that PR?

**Comment by noamd-legit** [comment]
@david-a-wheeler 
Removed the file here: https://github.com/ossf/wg-best-practices-os-developers/pull/508

**Review by balteravishay — APPROVED** [review]

**Pull request merged**